service/history/nDCActivityReplicator.go

```diff
@@ -202,63 +202,70 @@ func (r *nDCActivityReplicatorImpl) shouldApplySyncActivity(
 		if err != nil {
 			return false, err
 		}
+
 		lastLocalItem, err := currentVersionHistory.GetLastItem()
 		if err != nil {
 			return false, err
 		}
+
 		incomingVersionHistory := persistence.NewVersionHistoryFromThrift(incomingRawVersionHistory)
 		lastIncomingItem, err := incomingVersionHistory.GetLastItem()
 		if err != nil {
 			return false, err
 		}
-		if lastIncomingItem.GetVersion() < lastLocalItem.GetVersion() {
-			// the incoming branch will lose to the local branch
-			// discard this task
-			return false, nil
-		}
 
 		lcaItem, err := currentVersionHistory.FindLCAItem(incomingVersionHistory)
 		if err != nil {
 			return false, err
 		}
 
-		// version history matches and activity schedule ID appears in local version history
-		if currentVersionHistory.IsLCAAppendable(lcaItem) && scheduleID <= lastLocalItem.GetEventID() {
-			return true, nil
-		}
+		if currentVersionHistory.IsLCAAppendable(lcaItem) || incomingVersionHistory.IsLCAAppendable(lcaItem) {
+			// case 1: local version history is superset of incoming version history
+			// or incoming version history is superset of local version history
+			if scheduleID > lcaItem.GetEventID() {
+				// get event from the incoming cluster
+				var endEventID *int64
+				var endEventVersion *int64
+				if scheduleID < lastIncomingItem.GetEventID() {
+					// resend is exclusive-exclusive. Get the last event to be the schedule event
+					endEventID = common.Int64Ptr(scheduleID + 1)
+					endVersion, err := incomingVersionHistory.GetEventVersion(scheduleID + 1)
+					if err != nil {
+						return false, err
+					}
+					endEventVersion = common.Int64Ptr(endVersion)
+				}
 
-		// incoming branch will win the local branch
-		// resend the events with higher version
-		if scheduleID+1 <= lastIncomingItem.GetEventID() {
-			// according to the incoming version history, we can calculate the
-			// the end event ID & version to fetch from remote
-			endEventID := scheduleID + 1
-			endEventVersion, err := incomingVersionHistory.GetEventVersion(scheduleID + 1)
-			if err != nil {
-				return false, err
+				return false, newNDCRetryTaskErrorWithHint(
+					domainID,
+					workflowID,
+					runID,
+					common.Int64Ptr(lcaItem.GetEventID()),
+					common.Int64Ptr(lcaItem.GetVersion()),
+					endEventID,
+					endEventVersion,
+				)
+			}
+		} else {
+			// case 2: lcaItem is not the last item of incoming version history nor local version history
+			// indicates incoming version <> local version
```

Comment:
// case 2: incoming version history & local version history diverges

and plz move all description to the very top, like
```
// case 1. <desc>
// case 2. <desc>

if {
 // case 1
} else {
 // case 2
}